app/lib/shared/cached_value.dart

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:meta/meta.dart';
+import 'package:retry/retry.dart';
+
+final _logger = Logger('cached_value');
+
+typedef UpdateFn<T> = Future<T> Function();
+
+/// A locally cached and periodically updated value that can be accessed
+/// synchronously.
+class CachedValue<T> {
```

Comment:
Could we expand documentation to say something like:
```
/// Wraps a locally cached value that is periodically updated, and can be accessed synchronously.
///
/// **Example**:
/// ```dart
/// final foo = CacheValue(() => retry(() async {
///   return fetchFooFromSomewhere();
/// }));
///
/// // Access foo, notice that it's important to be able to handle cases
/// // where foo isn't available, because it haven't been loaded yet.
/// if (foo.isAvailable) {
///  print(foo.someMethod());
/// }
/// ```
///
/// The `update` function should do have it's own retry logic.
/// `CacheValue` will eventually retry the `update` function when the next scheduled update happens.
```

Reply:
We also need to determine a monitoring story here.

Should we maybe give `CacheValue` a few things during setup:
 * an `update` function which fetches a new value,
 * a `timeout` after which an old cached value becomes unavailable? is it ever preferable to return `null` if the cached value is really old?
 * a `name` that can be printed to logs and alerts,
 * an `interval` that says how frequently it should be updated,
 * Then `scheduleUpdates` might just be a  `startUpdates()` without any parameters.

This way, if you "goto references" for `CacheValue` constructor, then you can easily spot the options all cache values are configured with -- and it's easy to find the instance with a given name (say we get an alert that it's failing).

-------

From a monitoring perspective, all we can really do is maybe:
 * **warn** if `update` fails,
 * **shout** if `update` has failed so many times in a row that the cache value got older than `timeout` and, hence, has expired.

This way, for the youtube video listing we can say we have a `timeout` of 24 hours, and an update interval of 6 hours.
So if we fail to update 4 times in a row, then we'll get a **shout**, otherwise we only get warnings.

Similarly for name tracker... maybe we only care to shout if updating has failed for more than 30 minutes, assuming we update every 1-3 minutes.

Reply:
(let's have a chat about how far we want to go with this next week)